### PR TITLE
TCP Client socket support

### PR DESCRIPTION
**Pull request by wggihan**
## Purpose
This will introduce new struct to io package called 'Socket'. This client socket can be used to communicate with a remote TCP server.
Ref https://github.com/ballerina-lang/ballerina/issues/4720

## Goals
Communicate with a remote TCP server

## Release note
TCP Client socket support

## Automation tests
 - Unit tests 
   > 93%
 - Integration tests
   > N/A

## Samples
````
import ballerina.io;

function main (string[] args) {
    blob read;
    int size;
    string str;
    io:Socket client = io:openSocket("localhost", 9999, {});
    io:println(client.port);
    io:println(client.localPort);
    io:println(client.address);
    io:println(client.localAddress);
    io:ByteChannel channel = client.channel;
    string text = "Sample Text\n";
    blob content = text.toBlob("UTF-8");
    int count = channel.writeBytes(content, 0);
    read, size = channel.readBytes(12);
    str = read.toString("UTF-8");
    io:println(str);
    io:println(size);
    text = "Bye\n";
    count = channel.writeBytes(text.toBlob("UTF-8"), 0);
    read, size = channel.readBytes(4);    
    str = read.toString("UTF-8");
    io:println(str);
    io:println(size);
    client.closeSocket();
}
````

### CODE REVIEW

File: tests/ballerina-test/src/test/resources/test-src/io/clientsocketio.bal [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    io:ByteChannel channel = socket.channel;
+    blob readContent;
+    int readSize;
+    readContent, readSize = channel.readBytes(size);
+    return readContent, readSize;
+}
+
+function close(io:Socket socket) {
+    socket.closeSocket();
+}
```
Comment (pamod) [inline tests/ballerina-test/src/test/resources/test-src/io/clientsocketio.bal]: Shall we leave an empty space here ?

Comment (wggihan) [reply tests/ballerina-test/src/test/resources/test-src/io/clientsocketio.bal]: 46cd5053ef87419f9e68ead4cc40ddc161ccd086

File: stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/IOConstants.java [hunk 2]
```diff
@@ -42,4 +42,6 @@
      * The size of the buffer allocated for reading bytes from the channel (15~ KB).
      */
     public static final int CHANNEL_BUFFER_SIZE = 16384;
+
+    public static final String CLIENT_SOCKET_NAME = "clientSocket";
```
Comment (djkevincr) [inline stdlib/ballerina-builtin/src/main/java/org/ballerinalang/nativeimpl/io/IOConstants.java]: If possible, Shall we document this constant as other cases? Seems it is referred as key for native data for struct.

File: tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/io/socket/ClientSocketTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        BRunUtil.invoke(socketClient, "closeSocket");
+    }
+
+    @Test(dependsOnMethods = "testClosure", description = "Test connection open with properties")
+    public void testOpenWithProperties() {
+        int port = ThreadLocalRandom.current().nextInt(33000, 44000);
+        PackageInfo ioPackageInfo = socketClient.getProgFile().getPackageInfo("ballerina.io");
+        StructInfo socketProperties = ioPackageInfo.getStructInfo("SocketProperties");
+        BStruct propertyStruct = BLangVMStructs.createBStruct(socketProperties, port);
+        BValue[] args = {new BString("localhost"), new BInteger(port), propertyStruct};
```
Comment (Manuri) [inline tests/ballerina-test/src/test/java/org/ballerinalang/test/nativeimpl/functions/io/socket/ClientSocketTest.java]: Shall we fix formatting?

**Pull request merged**